### PR TITLE
stm32l1/pwm: PWM driver implementation for stm32l1

### PR DESCRIPTION
**Pull request by shady33**
The PWM peripheral implementation for the STM32L1 series.

### CODE REVIEW

**Comment by DipSwitch** [comment]
Sorry for the delay! Got myself an l1disovery, will validate this evening. (Both PR's)

File: boards/nucleo-l1/include/periph_conf.h [hunk 1]
```diff
@@ -213,6 +213,46 @@ static const timer_conf_t timer_config[] = {
...
+ */
+#define PWM_NUMOF           (2U)
+#define PWM_0_EN            1
+#define PWM_1_EN            1 
+#define PWM_MAX_CHANNELS    4
+
+/* PWM 0 device configuration */
+#define PWM_0_DEV           TIM3
+#define PWM_0_CHANNELS      4
+#define PWM_0_CLK           (168000000U)
```
Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: I doubt the source clock of the PWM runs at 168MHz ;) Guess this is copied from the F4.

File: boards/nucleo-l1/include/periph_conf.h [hunk 2]
```diff
@@ -213,6 +213,46 @@ static const timer_conf_t timer_config[] = {
...
+#define PWM_0_PIN_CH0       4
+#define PWM_0_PIN_CH1       5
+#define PWM_0_PIN_CH2       0
+#define PWM_0_PIN_CH3       1 
+#define PWM_0_PIN_AF        2
+
+/* PWM 1 device configuration */ 
+#define PWM_1_DEV           TIM4
+#define PWM_1_CHANNELS      4
+#define PWM_1_CLK           (84000000U)
```
Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: Also 84MHz is a little fast for an L1 :)

File: cpu/stm32l1/periph/pwm.c [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        case PWM_1:
+            PWM_1_CLKEN();
+            break;
+#endif
+    }
+}
+
+void pwm_poweroff(pwm_t dev)
+{
+    switch (dev) {
```
Comment (DipSwitch) [inline cpu/stm32l1/periph/pwm.c]: please use the new `periph_clk_dis`

File: cpu/stm32l1/periph/pwm.c [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+#if PWM_1_EN
+        case PWM_1:
+            PWM_1_DEV->CR1 &= ~(TIM_CR1_CEN);
+            break;
+#endif
+    }
+}
+
+void pwm_poweron(pwm_t dev)
+{
```
Comment (DipSwitch) [inline cpu/stm32l1/periph/pwm.c]: please use the new `periph_clk_en`

File: boards/nucleo-l1/include/periph_conf.h [hunk 5]
```diff
@@ -213,6 +213,46 @@ static const timer_conf_t timer_config[] = {
...
+/* PWM 0 device configuration */
+#define PWM_0_DEV           TIM3
+#define PWM_0_CHANNELS      4
+#define PWM_0_CLK           (168000000U)
+#define PWM_0_CLKEN()       (RCC->APB1ENR |= RCC_APB1ENR_TIM3EN)
+#define PWM_0_CLKDIS()      (RCC->APB1ENR &= ~RCC_APB1ENR_TIM3EN)
+/* PWM 0 pin configuration */
+#define PWM_0_PORT          GPIOB
+#define PWM_0_PORT_CLKEN()  (RCC->AHBENR |= RCC_AHBENR_GPIOBEN)
+#define PWM_0_PIN_CH0       4
```
Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: PB4 is used as defined as EXTI in the `nucleo-l1/include/periph_conf.h`

File: boards/nucleo-l1/include/periph_conf.h [hunk 6]
```diff
@@ -213,6 +213,46 @@ static const timer_conf_t timer_config[] = {
...
+/* PWM 1 device configuration */ 
+#define PWM_1_DEV           TIM4
+#define PWM_1_CHANNELS      4
+#define PWM_1_CLK           (84000000U)
+#define PWM_1_CLKEN()       (RCC->APB1ENR |= RCC_APB1ENR_TIM4EN)
+#define PWM_1_CLKDIS()      (RCC->APB1ENR &= ~RCC_APB1ENR_TIM4EN)
+/* PWM 1 pin configuration */
+#define PWM_1_PORT          GPIOB
+#define PWM_1_PORT_CLKEN()  (RCC->AHBENR |= RCC_AHBENR_GPIOBEN)
+#define PWM_1_PIN_CH0       6
```
Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: PB6 is used as defined as SPI CS line in `nucleo-l1/include/periph_conf.h`

File: boards/nucleo-l1/include/periph_conf.h (hunk 1 above)
Comment (shady33) [inline boards/nucleo-l1/include/periph_conf.h]: Sorry for that, overlooked the frequencies .. will change it to 32MHZ

File: boards/nucleo-l1/include/periph_conf.h [hunk 7]
```diff
@@ -213,6 +213,46 @@ static const timer_conf_t timer_config[] = {
...
+#define PWM_1_PIN_CH1       7 
+#define PWM_1_PIN_CH2       8
```
Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: PB8 is used as I2C SCL line `nucleo-l1/include/periph_conf.h`

File: boards/nucleo-l1/include/periph_conf.h [hunk 8]
```diff
@@ -213,6 +213,46 @@ static const timer_conf_t timer_config[] = {
...
+#define PWM_1_PIN_CH3       9
```
Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: PB8 is used as I2C SDA line `nucleo-l1/include/periph_conf.h`

File: boards/nucleo-l1/include/periph_conf.h (hunk 5 above)
Comment (shady33) [inline boards/nucleo-l1/include/periph_conf.h]: Should i use different pins or is it better to remove these definitions?

Comment (DipSwitch) [inline boards/nucleo-l1/include/periph_conf.h]: I think if we remove PB4 from the EXTI defines we could remove all the PWM1 defines and have a full functioning PWM_0_DEV :)

@thomaseichinger Is the EXTI line PB4 used for something? Or can it's definition be removed to make place for PWM?

**Comment by kYc0o** [comment]
Feature freeze -> postponed.

**Comment by PeterKietzmann** [comment]
@A-Paul you are interested in PWMs :-). Would you like to take this PR over after the feature freeze? If so, yust self-assign. We have one stm32l1 nucleo board.

**Comment by miri64** [comment]
Postponed due to Feature Freeze.

**Comment by PeterKietzmann** [comment]
Obsolete with #6186.

**Pull request closed without merging**